tests/test_read.py

```diff
@@ -319,3 +321,22 @@ def test_read_out_mask(path_rgb_byte_tif, out):
     with rasterio.open(path_rgb_byte_tif) as src:
         with pytest.raises(ValueError):
             src.read(indexes=[2], out=out)
+
+
+def test_chained_io_errors(path_rgb_byte_tif):
+    """Get chained exceptions."""
+    with rasterio.open("tests/data/corrupt.tif") as src:
+        with pytest.raises(RasterioIOError) as excinfo:
+            src.read()
+        assert isinstance(excinfo.value.__context__, CPLE_AppDefinedError)
+        assert excinfo.value.__context__.errmsg.startswith(
+            "tests/data/corrupt.tif, band 1: IReadBlock failed at X offset 1, Y offset 1"
```

Comment:
testing exact error messages looks fragile to me. You could potentially just test they are not empty and different between that instance and the one 3 lines below